bpf/nat64.c

```diff
@@ -43,127 +43,174 @@
+			// ordered differently and have different sizes. For checksum
+			// calculation, this does not matter - all bytes of
+			// pseudo-header apart from IP addresses contribute the
+			// same value to checksum (first step of calculation is
+			// summing all bytes, zeroes does not matter),
+			// meaning we only need to run bpf_csum_diff over IP
+			// addresses instead of the entire pseudo-header.
+			//
+			// Last neat piece of info that makes it a one-liner is that both
+			// ipv6hdr and iphdr structs have src and dst addresses
+			// next to each other in memory. That means we can
+			// calculate checksum difference with one bpf_csum_diff
+			// call using 2 * size of IP address.
+			l4_csum_diff = bpf_csum_diff((void *)&(ip6->saddr), 2*sizeof(struct in6_addr), (void *)&(ip4.saddr), 2*sizeof(__u32), 0);
+	}
 
 	// Packet mutations begin - point of no return, but if this first modification fails
 	// the packet is probably still pristine, so let clatd handle it.
-	if (bpf_skb_change_proto(skb, bpf_htons(ETH_P_IP), 0))
+	// This also takes care of resizing socket buffer to handle different IP
+	// header size.
+	if (bpf_skb_change_proto(skb, bpf_htons(ETH_P_IP), 0)) {
+		bpf_printk("NAT64 packet forwarded: bpf_skb_change_proto failed");
 		return TC_ACT_OK;
+	}
 
-	// recalculate protocol checksums
-	// UDP checksum is optional for IPv4
-	if (ip.protocol == IPPROTO_UDP) {
-		__u16 new_csum = 0;
-		bpf_skb_store_bytes(skb, UDP_CSUM_OFF, &new_csum, sizeof(new_csum), 0);
+	// Update L4 checksum using the checksum difference we calculated before.
+	int ret = 0;
+	switch (ip4.protocol) {
+		case IPPROTO_UDP:
+			ret = bpf_l4_csum_replace(skb, UDP_CSUM_OFF, 0, l4_csum_diff, BPF_F_PSEUDO_HDR);
+			break;
+		case IPPROTO_TCP:
+			ret = bpf_l4_csum_replace(skb, TCP_CSUM_OFF, 0, l4_csum_diff, BPF_F_PSEUDO_HDR);
+			break;
 	}
-	// TODO: recalculate TCP checksum since we can not depend on checksum neutrality
-	if (ip.protocol == IPPROTO_TCP) {
-		__u16 new_csum = 0;
-		bpf_skb_store_bytes(skb, TCP_CSUM_OFF, &new_csum, sizeof(new_csum), 0);
+
+	// If true, updating packet's UDP / TCP checksum failed.
+	if (ret < 0) {
+		bpf_printk("NAT64 packet dropped: L4 checksum update failed");
+		return TC_ACT_SHOT;
 	}
 
 	// bpf_skb_change_proto() invalidates all pointers - reload them.
 	data = (void*)(long)skb->data;
 	data_end = (void*)(long)skb->data_end;
 	// I cannot think of any valid way for this error condition to trigger, however I do
 	// believe the explicit check is required to keep the in kernel ebpf verifier happy.
-	if (data + l2_header_size + sizeof(struct iphdr) > data_end)
+	if (data + sizeof(struct ethhdr) + sizeof(struct iphdr) > data_end)
 		return TC_ACT_SHOT;
 
-	struct ethhdr* new_eth = data;
-	// Copy over the updated ethernet header
-	*new_eth = eth2;
+	// Copy over the old ethernet header with updated ethertype.
+	ret = bpf_skb_store_bytes(skb, 0, &old_eth, sizeof(struct ethhdr), 0);
```

Comment:
reading the docs https://man7.org/linux/man-pages/man7/bpf-helpers.7.html it seems this is safer than the previous approach that manipulated directly the skb